src/server/pkg/dlock/dlock.go

```diff
@@ -0,0 +1,74 @@
+// Package dlock implements a distributed lock on top of etcd.
+package dlock
+
+import (
+	"context"
+
+	etcd "github.com/coreos/etcd/clientv3"
+	"github.com/coreos/etcd/clientv3/concurrency"
+)
+
+// DLock is a handle to a distributed lock.
+type DLock interface {
+	// Unlock releases the distributed lock.
+	Unlock() error
+	// Context returns a context that's cancelled if the lock has been
+	// released for any reason, e.g. if the node holding the lock has
+	// disconnected from etcd.
+	// After you've acquire the lock, you should use this context in any
+	// subsequent blocking requests, so that if you lose the lock, the
+	// requests get cancelled correctly.
+	Context() context.Context
+}
```

Comment:
If it's not too much trouble I'd like to see this with an API more similar `sync.Mutex` with a `Lock` and `Unlock` method. Actually I'm wondering if maybe it'd be simpler to just define an interface with `Lock` and `Unlock` that `concurrency.Mutex` satisfies and return it directly.